### DISCUSSION TITLE
Use GLTFast to implement 3Dtiles in Unity

### DISCUSSION THREAD

**Discussion by hcshen** [atteneder/glTFast#429] [General]
Currently I rely on GLTFast to implement 3Dtiles in Unity, and load around 50G of data to show 60 frames. The plugin works well, but currently it is difficult to use multi-threading for instantiation. Do you have any plans to support the 3Dtiles specification？

**Comment by hcshen** [comment]
Sorry, my english is poor

**Comment by atteneder** [comment]
Hi,

I don't have a deep knowledge about 3D tiles and what implications it has on glTF loading.

What is your main pain point? Are you saying loading times or frame rates are a problem and you suspect it could be improved by making instantiation threaded?

There are plans to make instantiation async. I assume it will improve the frame rate, but I don't expect major performance boost. Instantiation mainly is about firing lots of Unity API calls which are all bound to the main thread and thus don't allow for much threading improvements.

The experimental DOTS support on the other hand has the potential to be a lot faster, but there's a reason it's still experimental.

**Reply by hcshen** [reply]
Thanks,I am currently working on loading 3DTils data in Unity platform. My pain point is that loading a large amount of 3DTils data may cause a large amount of IO per frame, as well as a large number of GLTF objects instantiation and unloading. Now I have achieved the loading effect of 60 frames, but when the camera moves to refresh 3DTiles, there are frame drops and lag. I don't know if the gltf instantiation will improve the Frame dropping phenomenon if it supports multithreading, here are 3 dtiles specification: https://github.com/CesiumGS/3d-tiles

**Reply by hcshen** [reply]
Through Profiler processing here will cost around 200 ms, frame drop phenomenon is supposed to be here to: Profiler. BeginSample (" LoadAccessorData. IndicesMatrixJob "),

**Reply by atteneder** [reply]
Ok, so the problem is an instable frame rate. Got it.

Thanks for sharing the profiling result.
This particular region is mainly doing:
- Allocating destination buffers
- Starting (threaded) jobs to fill them
in a loop. It should defer execution to the next frame, if the load is very heavy, so I'm surprised that it stalls your app for 200ms. I could dig deeper if you could provide some glTFs to profile with.

The upcoming async instantiation probably improves this, but probably not much by threading.

**Comment by matrem** [comment]
Hello,
I guess unity.gis.streaming-framework as a 3D tiles implementation based on GLTFast : https://github.com/Unity-Technologies/com.unity.gis.streaming-framework.
I did not test it, but I have to try to know their performances.

**Reply by hcshen** [reply]
Ok, Thank you very much! I will to test it
